forem/forem
# Provide way to resend comfirmation email via log-in page

## What type of PR is this? (check all applicable)
Feature and Bug fix

## Description
In the self-hosted Forem instance context, admins are allowed to proceed to set up the app without an SMTP credentials. When they do enable it, our app would automatically log all users out until they "confirm their email". The problem is that they would not be able to do that because the app never sent it and would see the following error message. 

 <img width="300"  src="https://user-images.githubusercontent.com/15793250/123972310-bb888c80-d988-11eb-8f34-744ca6f76407.png">

This provides a solution to this problem in 2 parts
1. We will wave Admin's need to confirm their email with `.skip_confirmation!`. It doesn't make sense to log out the only person in the app that can make changes to the app and I'm making the big assumption that they have no malicious intent and have the right email address.
2. The error message is updated with a link to "resend confirmation email" (see below). Which would redirect them to our updated `/confirm-email` page. The functionality of that page is not modified and will work the same way for users signed up via email.

This is initially created as a stop-gap solution but I feel like this could be the ideal solution. The other idea that few came up with was a background task that would automatically send out confirmation emails when the SMTP credential is applied. I don't think that's a good idea because an unexpected confirmation email they didn't ask for is not a good surprise.

## Related Tickets & Documents
n/a

## QA Instructions, Screenshots, Recordings
For proper testing, this QA instruction will ask you to reset your database. It's not difficult to QA this, the instruction is verbose.

The setup phase
1. Mimic new Forem instance by adding the following two to your .env 
    ```
   FOREM_OWNER_SECRET="secret"
   MODE="STARTER"
   ```
2. Please reset your database so you can mimic admin sign-up flow with `bin/rails db:reset && bin/rails db:setup`
3. If you have SMTP enabled, please disable it/remove the env variable associated with it. 
4. Start up the server, navigate to the homepage and sign up for an account.
5. Head to `/admin/config` and enable email sign up under the authentication section (disable all other options if it won't you update).
6. On a different browser or an incognito tab, sign up for a standard user account
7. Stop the server. We're going to pretend SMTP is enabled by adding the following to .env
    ```
    SENDGRID_API_KEY="something"
    ```

The testing phase
1. Startup the server and navigate to the homepage
2. As the admin, you should not be logged out and should be able to navigate just fine.
3. As a user, you should be logged out and see the following message.
   <img width="300"  src="https://user-images.githubusercontent.com/15793250/123973221-95172100-d989-11eb-82de-263b2aed4baf.png">
4. Click on the resend confirmation link and you should redirect to this page.
   <img width="300"  src="https://user-images.githubusercontent.com/15793250/123973336-ad873b80-d989-11eb-88ee-91473ae70049.png">
5. Enter the email of your user account and click submit. It should succeed without issue.
6. Navigate to http://localhost:3000/rails/mailers/devise/mailer/confirmation_instructions and you should see a confirmation email exists.
7. That's it!


### UI accessibility concerns?
Nope!

## Added/updated tests?
- I did not add any. At its core, I added a link to an error message that redirects users to an existing feature. I don't think that warrants a spec.

## [Forem core team only] How will this change be communicated?
- I will share this change internally with the appropriate teams


## [optional] Are there any post-deployment tasks we need to perform?
n/a
